### ISSUE TITLE
Outdated resource is passed to the reconcile method

### ISSUE THREAD

**Issue by bachmanity1** [java-operator-sdk/java-operator-sdk#1906]
I have observed that after successfully updating the status of a resource, there are instances where the subsequent invocation of the `reconcile` method may still have an outdated resource with an un-updated status. Is this an intended behavior or a bug?

Since my `reconcile` method relies on the status of the resource, having an outdated resource with an un-updated status could result in unexpected behavior.

**Comment by csviri** [comment]
Hi @bachmanity1 ,

yes, this is expected. There are multiple aspects of this:

1. by definition the status should not be used to persist state, in terms that there should not be made any decision based on actual status during the reconciliation. Status should be rather the last observed state. 

2. since status update usually is a patch (without optimistic locking this), in general this is not properly solvable to cache the resource (some concurrency issues, won't go into details) even if `UpdateControl` is used. Some teams actually use status as storing states, but then doing some tricks to explicitly cache it into memory and make sure it is always the latest version is used.

3. Some Kubernetes backup tools don't backup status by default (if remember correctly for example https://velero.io/ ). This is an other reason, not to use status for persisting state. Although it can be explicitly marked to a resource to also backup the status.

So in short status should not be used to store state, but some operators are implemented like that, but this is not supported out of the box by the framework.

The advised solution is to use a ConfigMap/Secret or dedicated Custom Resource. Note that even that is a bit tricky, but we support it in dependent resources explicitly, you can use that or implement it using low level APIs yourself:
https://javaoperatorsdk.io/docs/dependent-resources#external-state-tracking-dependent-resources

**Comment by csviri** [comment]
In general it is possible to have outdated resource from Informer or any Event Sources, the connections can drop or lagging, or just havin the reconciliation started parallel with an actually happening updated in the API Server. But Operator Pattern is designed to cover these cases.

**Comment by bachmanity1** [comment]
Ok, thanks for the clarification! It seems that I need to reconsider the design of my reconciler. 😢
